x/qgb/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,195 @@
+			EthAddress:   oc.evmAddress.Hex(),
+			Nonce:        valset.Nonce,
+			Signature:    ethcmn.Bytes2Hex(signature),
+		}
+
+		err = oc.broadcastTx(ctx, msg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (oc *orchestrator) orchestrateDataCommitments(ctx context.Context) error {
+	queryClient := types.NewQueryClient(oc.qgbRPC)
+
+	resp, err := queryClient.Params(ctx, &types.QueryParamsRequest{})
+	if err != nil {
+		return err
+	}
+
+	params := resp.Params
+
+	results, err := oc.tendermintRPC.Subscribe(ctx, "height", coretypes.EventQueryNewBlockHeader.String())
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case ev := <-results:
+			oc.processDataCommitmentEvents(ctx, params.DataCommitmentWindow, ev)
+		}
+	}
+}
+
+func (oc *orchestrator) processDataCommitmentEvents(ctx context.Context, window uint64, ev rpctypes.ResultEvent) error {
+	eventDataHeader := ev.Data.(coretypes.EventDataNewBlockHeader)
+	height := eventDataHeader.Header.Height
+	// todo: refactor to ensure that no ranges of blocks are missed if the
+	// parameters are changed
+	if height%int64(window) != 0 {
+		return nil
+	}
+
+	// TODO: calculate start height some other way that can handle changes
+	// in the data window param
+	startHeight := height - int64(window)
+	endHeight := height
+
+	// create and send the data commitment
+	dcResp, err := oc.tendermintRPC.DataCommitment(
+		ctx,
+		fmt.Sprintf("block.height >= %d AND block.height <= %d",
+			startHeight,
+			endHeight,
+		),
+	)
+	if err != nil {
+		return err
+	}
+
+	nonce, err := oc.wrapper.StateLastDataRootTupleRootNonce(&bind.CallOpts{})
+	if err != nil {
+		return err
+	}
+
+	nonce.Add(nonce, big.NewInt(1))
+
+	dataRootHash := types.DataCommitmentTupleRootSignBytes(oc.bridgeID, nonce, dcResp.DataCommitment)
+
+	dcSig, err := oc.personalSignerFn(oc.evmAddress, dataRootHash.Bytes())
+	if err != nil {
+		return err
+	}
+
+	msg := &types.MsgDataCommitmentConfirm{
+		EthAddress:       oc.evmAddress.String(),
+		Commitment:       dcResp.DataCommitment.String(),
+		BeginBlock:       startHeight,
+		EndBlock:         endHeight,
+		ValidatorAddress: oc.signer.GetSignerInfo().GetAddress().String(),
+		Signature:        ethcmn.Bytes2Hex(dcSig),
+	}
+
+	return oc.broadcastTx(ctx, msg)
+}
+
+func (oc *orchestrator) broadcastTx(ctx context.Context, msg sdk.Msg) error {
+	err := oc.signer.QueryAccountNumber(ctx, oc.qgbRPC)
+	if err != nil {
+		return err
+	}
+
+	// TODO: update this api via https://github.com/celestiaorg/celestia-app/pull/187/commits/37f96d9af30011736a3e6048bbb35bad6f5b795c
+	tx, err := oc.signer.BuildSignedTx(oc.signer.NewTxBuilder(), msg)
+	if err != nil {
+		return err
+	}
+
+	rawTx, err := oc.signer.EncodeTx(tx)
+	if err != nil {
+		return err
+	}
+
+	resp, err := paytypes.BroadcastTx(ctx, oc.qgbRPC, 1, rawTx)
+	if err != nil {
+		return err
+	}
+
+	if resp.TxResponse.Code != 0 {
```

Comment:
We need some retry mechanism. Since, the data commitments confirms will be consecutive. If enough fo these fail, then it would create a gap

Reply:
I agree! This is something that I figured that we could dig into and think about more after merging imho

We should definitely write an issue about this before merging

Reply:
this issue is enough I guess: https://github.com/celestiaorg/celestia-app/issues/260